### DISCUSSION TITLE
Debian Installation help

### DISCUSSION THREAD

**Discussion by Uzzi** [uhppoted/uhppoted#11] [General] [answered]
Hi, I've a Debian 11 vm an I have to install all.
Please there is a step by step howto?
Thank you

Regards

**Comment by uhppoted** [comment, marked as answer]
Hi,

If you really want **everything** then you'll have to clone the [uhppoted](https://github.com/uhppoted/uhppoted) repository and build it from scratch - the instructions are [here](https://github.com/uhppoted/uhppoted#building-from-source).

However, I suspect you probably just want the executables for Debian - the easiest way to get those is to download the latest release from the [Releases](https://github.com/uhppoted/uhppoted/releases) page (e.g. [uhppoted_v0.7.3-linux.tar.gz](https://github.com/uhppoted/uhppoted/releases/download/v0.7.3/uhppoted_v0.7.3-linux.tar.gz) and extract it to a folder of your choice. 

For example, to install the executables as user _uhppoted_ in folder _/opt/uhppoted_:

```
# Create user uhppoted
sudo adduser uhppoted
sudo usermod -aG sudo uhppoted

# Create /etc/uhppoted folder for configuration information
sudo mkdir -p /etc/uhppoted
sudo chown -R uhppoted:uhppoted /etc/uhppoted

# Use folder /opt/uhppoted
sudo mkdir -p /opt/uhppoted
sudo chown -R uhppoted:uhppoted /opt/uhppoted
sudo su uhppoted
cd /opt/uhppoted

# Download and extract tar file
wget https://github.com/uhppoted/uhppoted/releases/download/v0.7.3/uhppoted_v0.7.3-linux.tar.gz
tar xvzf uhppoted_v0.7.3-linux.tar.gz

# (optional) Create symbolic links in /usr/local/bin
sudo ln -s /opt/uhppoted/uhppoted_v0.7.3/uhppote-cli /usr/local/bin/uhppote-cli

# Test
uhppote-cli --debug get-devices
```

Something to be aware of:

The UHPPOTE controllers use a UDP based protocol which means it's going to be more complicated than you think to use a VM because UDP broadcast packets are restricted to the local network segment. Basically:
- any interposing firewalls, bridged networks, etc will need to be configured to allow UDP packets through on port 60000
- you will need an _/etc/uhppoted/uhppoted.conf_ file configured with the controller IP addresses, e.g.:
```
...
# DEVICES
UT0311-L0x.405419896.name = Controller1
UT0311-L0x.405419896.address = 192.168.1.100:60000
...
```

hth :-)

**Comment by uhppoted** [comment]
(BTW, the _uhppoted-httpd_ and _uhppoted-tunnel_ components aren't included in the 0.7.3 release - they're due for release sometime in the next couple of days)

**Comment by Uzzi** [comment]
Thank you so much!
Now I'm waiting uhppoted-httpd and  uhppoted-tunnel components.
My vm in on the same subnet of controlers!

**Comment by Uzzi** [comment]
If I need all components I 've to install go and nodejs, it it true?

**Reply by uhppoted** [reply]
Yup, you'll need Go for everything apart from the _uhppoted-nodejs_ and _node-red-contrib-uhppoted_ modules.

NodeJS is obviously required for the _uhpotted-nodejs_ module, but it's also needed for building _uhppoted-httpd_ and _uhppoted-tunnel_ because they both use _eslint_ for Javascript syntax/semantic checking and formatting.

BTW, _uhppoted-httpd_ and _uhppoted-tunnel_ are just about good to go (mostly it's just some additional documentation that's needed) so you should be able to build them by cloning the repos and following the _Building from source_ instructions

**Comment by Uzzi** [comment]
sorry, why this?
Clone in '/opt/uhppoted/uhppoted-dll' in corso...
git@github.com: Permission denied (publickey).

git clone --recurse-submodules https://github.com/uhppoted/uhppoted.git in my debian stops and it doen't complete processes

**Comment by uhppoted** [comment]
github recently moved to requiring secure _git_ URLs which seems to have resulted in a mix of _git://_ and _https://_ URLs in the _.gitmodules_ file. I'll fix it a bit later but for now just edit the .gitmodules file and replace the _git_ URLs with the _https_ equivalent:
```
[submodule "uhppote-core"]
    path = uhppote-core
    url = https://github.com/uhppoted/uhppote-core.git
[submodule "uhppoted-lib"]
    path = uhppoted-lib
    url = https://github.com/uhppoted/uhppoted-lib.git
[submodule "uhppote-simulator"]
    path = uhppote-simulator
    url = https://github.com/uhppoted/uhppote-simulator.git
[submodule "uhppote-cli"]
    path = uhppote-cli
    url = https://github.com/uhppoted/uhppote-cli.git
[submodule "uhppoted-rest"]
    path = uhppoted-rest
    url = https://github.com/uhppoted/uhppoted-rest.git
[submodule "uhppoted-mqtt"]
    path = uhppoted-mqtt
    url = https://github.com/uhppoted/uhppoted-mqtt.git
[submodule "uhppoted-app-sheets"]
    path = uhppoted-app-sheets
    url = https://github.com/uhppoted/uhppoted-app-sheets.git
[submodule "uhppoted-app-s3"]
    path = uhppoted-app-s3
    url = https://github.com/uhppoted/uhppoted-app-s3.git
[submodule "node-red-contrib-uhppoted"]
    path = node-red-contrib-uhppoted
    url = https://github.com/uhppoted/node-red-contrib-uhppoted.git
[submodule "uhppoted-app-wild-apricot"]
    path = uhppoted-app-wild-apricot
    url = https://github.com/uhppoted/uhppoted-app-wild-apricot
[submodule "uhppoted-httpd"]
    path = uhppoted-httpd
    url = https://github.com/uhppoted/uhppoted-httpd
[submodule "uhppoted-nodejs"]
    path = uhppoted-nodejs
    url = https://github.com/uhppoted/uhppoted-nodejs.git
[submodule "uhppoted-dll"]
    path = uhppoted-dll
    url = https://github.com:uhppoted/uhppoted-dll.git
[submodule "uhppoted-tunnel"]
    path = uhppoted-tunnel
    url = https://github.com/uhppoted/uhppoted-tunnel.git
```

**Comment by Uzzi** [comment]
I would like to develop a webinterface+db for card-users organization, 2 controllers(2doors) permissions, allow and deny log store ecc ecc.
Can I develop a PHP+mysql app that use uhppote-rest to dialogate with controllers? 
what's your advice?

**Comment by uhppoted** [comment]
There are so many options :-). For example:

- a pure PHP implementation using [carbonsphere's](https://github.com/carbonsphere/UHPPOTE) implementation as a base. It will be a fair amount of work but ultimately simpler because it's all one language and system.
- PHP does have an FFI (foreign function interface) so you could consider using the [DLL](https://github.com/uhppoted/uhppoted-dll) but that would require wrapping the DLL functions in PHP equivalents and TBH that isn't trivial
- the REST server is an option too, though personally I think it's probably overkill and unnecessarily complicated for this particular use case.
- use [NodeRED](https://nodered.org/) which does have a built-in MySQL connector

If it were me I would probably start out by:
1. Developing the UI and MySQL bits
2. Exporting the card permissions as an access control list (ACL) file from MySQL
3. Using either [uhppote-cli](https://github.com/uhppoted/uhppote-cli) or [uhppoted-app-s3](https://github.com/uhppoted/uhppoted-app-s3) to merge the ACL file to the controller. It's easy enough to run a bash script from PHP but a _cron_ job or a file watcher are other alternatives.

Once that's all working you could think about replacing the ACL file with something a bit more integrated.

**Comment by Uzzi** [comment]
In your opinion, it si better uhppoted system on Proxmox VM or on Rasperry PI 4?

**Comment by uhppoted** [comment]
I think the answer to that is "_whatever works for you_".

Personally I tend to use _Raspberry Pis_ (or occasionally small fanless embedded boards) because I find it simpler and easier:
- I prefer small dedicated systems that I can lock in a panel or cupboard and forget about
- If it fails it's only one subsystem that fails
- It's low cost (especially if you use Raspberry Pi 3Bs - and I haven't tried it but I suspect it's quite possible to run on a _Pi Zero_)
- You can have redundant/backup systems without taking out a second mortgage on your home
- And ... well, I just really, really **like** Raspberry Pis :-). They're solid, well-engineered, affordable and based on the ARM CPUs.

The one big disadvantage of the Raspberry Pi is the SD card which has a limited lifetime and is known to fail erratically (i.e. don't use cheap ones or use an SSD).  I haven't had one fail yet but TBH it's just a matter of time...

I'd be curious to know what you consider to be the advantages/disadvantages of using _Proxmox_?

**Comment by uhppoted** [comment]
(for interest only)

[Proxmox vs FreeBSD: Which Virtualization Host Performs Better?](https://it-notes.dragas.net/2024/06/10/proxmox-vs-freebsd-which-virtualization-host-performs-better)
